News/Sceen/Main/ViewModel/NewsViewModel.swift

```diff
@@ -0,0 +1,60 @@
+//
+//  NewsViewModel.swift
+//  News
+//
+//  Created by Mohamed Elboraey on 15/05/2025.
+//
+
+import Foundation
+import Combine
+
+typealias NewsViewModelProtocol = NewsViewModelInput & NewsViewModelOutput
+
+protocol NewsViewModelInput {
+    func getData(from url: String) async
+}
+
+protocol NewsViewModelOutput {
+    var newsData: NewsData? { get }
+    var error: String? { get }
+    var newsDataPublisher: AnyPublisher<NewsData?, Never> { get }
+    var errorPublisher: AnyPublisher<String?, Never> { get }
+    //    var onDataUpdate: (() -> Void)? { get set }
+    //    var onError: ((String) -> Void)? { get set }
+}
+
+final class NewsViewModel: NewsViewModelProtocol {
+    var newsDataPublisher:  AnyPublisher<NewsData?, Never> {$newsData.eraseToAnyPublisher()}
+    
+    var errorPublisher: AnyPublisher<String?, Never> {$error.eraseToAnyPublisher()}
+    
+//    var onDataUpdate: (() -> Void)?
+//    var onError: ((String) -> Void)?
+    
+    @Published var newsData: NewsData?
+    @Published var error: String?
+    
+    private let useCase: NewsUseCaseProtocol
+    
+    init(useCase: NewsUseCaseProtocol = NewsUseCase()) {
+        self.useCase = useCase
+    }
+}
+
+extension NewsViewModel {
+    func getData(from url: String) async {
+        do {
+            let data = try await useCase.fetchNews(url)
+            await MainActor.run {
+                self.newsData = data
+//                onDataUpdate?()
+            }
+        } catch {
+            await MainActor.run {
+                self.error = error.localizedDescription
+//                onError?(self.error ?? "")
+            }
+        }
+    }
+    
```

Comment:
I would suggest the view model to be marked as @MainActor instead. It will ensure all the properties are updated on the main actor.

```swift
@MainActor
final class NewsViewModel: NewsViewModelProtocol {
```

Then this should be 
```suggestion
extension NewsViewModel {
    func getData(from url: String) async {
        do {
            let data = try await useCase.fetchNews(url)
            newsData = data
        } catch {
            error = error.localizedDescription
        }
    }
    
```
As it should be running on the main actore already.